### DISCUSSION TITLE
Set up override for subject lines for user invite & password recovery mails

### DISCUSSION THREAD

**Discussion by anneQui** [directus/directus#20732] [Feature Requests]
### Summary

Hi guys,

Directus has a great system which provides user invites & password recovery features straight out of the box.
However the subjects for both mails are not modifiable currently.

There has been a demand from multiple users to set up a custom subject for the inviteUser & requestPasswordReset methods.
See:
 - https://github.com/directus/directus/discussions/17727 >Allow custom subject lines for new user invites
 - https://github.com/directus/directus/pull/7372 > Add Subject to password request 
 - https://github.com/directus/directus/discussions/17485 > Translatable API
 - https://github.com/directus/directus/pull/7412 > Create two new environment variables for subject in email 

There is already a system set up in the requests to pass a subject in the request to the API, would it be possible to use it while the API translation is being set up ?

### Basic Example

For the user invite, a call to the invite API could simply be modified from this:

```
await api.post('/users/invite', {
	email: emailsParsed,
	role: selectedRole,
});
```
to:
```
await api.post('/users/invite', {
	email: emailsParsed,
	role: selectedRole,
        subject: 'My custom subject line',
});
```

### Motivation

**Customization:**
Not the main motivator but developpers may want to use the site's name, etc or customize the e-mail.

**I18n:** 
The issue is a main issue for international users. **The e-mails sent by these two systems have set titles in English, making internationalization impossible for the emails.**

Here's a dramatic recreation:
![illustration_directus](https://github.com/directus/directus/assets/140420708/f0b1824f-cc70-432f-9866-7265521efc53)

### Detailed Design

As Jason mentioned in his[ previous discussion](https://github.com/directus/directus/discussions/17727) the system already exists and just needs to be implemented.

### Must Have:
Custom email subjects for new user invites

### Drawbacks

None.
The subject can be treated as an optional field, allowing for backwards compatibility. The existing implementation of the user invite feature can remain unchanged.

### Alternatives

Mentioned earlier here https://github.com/directus/directus/discussions/17485
The idea was to set up a translation for the API which is a larger project. If there is a status on this, I would gladly have a look :)

### Adoption Strategy

No impact on other project as far as I know.
Documentation should be updated to explain the subject can be customized, otherwise not much

**Comment by anneQui** [comment]
I'm tagging the people who previously created discussions on the subject as well to see if there's still interest on their side:
- @AndreyKindin 
- @JonathanSchndr 
- @JasonAGross

**Comment by dmsolutionz** [comment]
This old PR https://github.com/directus/directus/pull/17728 was closed due to no "Heya! Thank you for your contribution. Unfortunately, we can't accept PRs for improvements/features that don't have a prior accepted Discussion and associated Issue" I would say this 3 month old discussion and associated issue are more than enough to go ahead with the change, fairly low impact in terms of implementation.

If someone else needs to open a new PR, I can do this with the changes.

**Comment by kevinmamaqi** [comment]
What is the status of this?

**Comment by Blanco237** [comment]
Directus team what is going on with this? There have been multiple request and discussions on this topic over the years but you keep closing them down because the original poster didn't knock at your door consistently for 90days straight.
Worse off, a PR was opened which solves the issue without affecting the operation of the email API in any way, `4 lines` of code changed, **4 lines!!**, and yet the PR was closed and it's been 18 months since then.
Want to know what the most annoying part is, those discussions that you claim haven't received the 15 upvotes have since then received those 15 upvotes and more...

Maintainers, for the love of God please, reopen and merge the PR that fixes this issue.
@azrikahar @rijkvanzanten

**Reply by JonathanSchndr** [reply]
i totally understand that this issue is causing resentment and i also hope that there will soon be a solution or at least a priority in the first place.

**Reply by JonathanSchndr** [reply]
... I remember that someone from the Directus team once had an example to map this via a flow and send the mail with it. But I can't remember where I saw it.

**Reply by Blanco237** [reply]
I actually did this on our application. I should probably write a blog or make a video on it...

**Comment by dforce2055** [comment]
Any updates?

**Reply by Blanco237** [reply]
None yet brother...
But I found a sort of work around that you can use.
By triggering the email manually with a flow and setting the subject line.
I'll write a blog/YouTube video on it soon

**Reply by dforce2055** [reply]
> None yet brother... But I found a sort of work around that you can use. By triggering the email manually with a flow and setting the subject line. I'll write a blog/YouTube video on it soon

Sounds good, I look forward to updates from your side!
In the meantime, here are some workarrounds I'm thinking about this issue

This is the problematic line

**https://github.com/directus/directus/blob/189fcb24f7bb12118c89d04639a971d4afe588e2/api/src/services/users.ts#L359C1-L359C58**

```typescript:
const subjectLine = subject ?? "You've been invited";
```

Just as we have environment variables to define

```
PASSWORD_RESET_URL_ALLOW_LIST
USER_INVITE_URL_ALLOW_LIST
USER_REGISTER_URL_ALLOW_LIST

```
We could think of defining
```
PASSWORD_RESET_SUBJECT
USER_INVITE_SUBJECT
USER_REGISTER_SUBJECT
```

And use it in the code

```const subjectLine = subject ?? env['PASSWORD_RESET_SUBJECT'] as string; ```


Another idea I have is to intercept the mail with mailgun before it is sent to the user and change the subject. I'm not sure if this is possible, but I will try it.

**Reply by phazonoverload** [reply]
You can intercept the email.send event with flows or a hook and change content there. Not an elegant fix but workable for lack of this feature.

**Comment by woeden** [comment]
Any updates ? 🙏 

I use this hook but it would be so much more practical to set environment variables.

```js
import { defineHook } from '@directus/extensions-sdk';

export default defineHook(({ filter }, { services, logger, getSchema, env }) => {
	filter('email.send', async (input: any) => {
		const templateName = input.template.name;
		switch (templateName){
			case 'user-invitation':
				input.subject = 'Vous a avez été invité à nous rejoindre'
				break;
			case 'password-reset':
				input.subject = 'Votre lien de réinitialisation de mot de passe'
				break;
		}
		return input;
	});
});
```

**Comment by podlomar** [comment]
Hi Directus team, I stumbled upon this too. I am building an application that is not in English and having subject lines of the automatic e-mails in my app in English looks unprofessional and "half-baked". Can you please fix this, pretty pretty please?

**Reply by highvibesonly** [reply]
same here...
